pkg/nsx/services/subnetport/subnetport.go

```diff
@@ -0,0 +1,301 @@
+		log.Error(err, "check and update NSX subnet port state failed, would retry exponentially", "subnetport", obj.UID)
+		return err
+	}
+	log.Info("successfully created or updated subnetport", "subnetport", obj.UID)
+	return nil
+}
+
+// CheckSubnetPortState will check the port realized status then get the port state to prepare the CR status.
+func (service *SubnetPortService) CheckSubnetPortState(obj *v1alpha1.SubnetPort) error {
+	nsxSubnetPort := service.SubnetPortStore.GetByKey(string(obj.UID))
+	if nsxSubnetPort == nil {
+		return errors.New("failed to get subnet port from store")
+	}
+	realizeService := realizestate.InitializeRealizeState(service.Service)
+	if err := realizeService.CheckRealizeState(retry.DefaultRetry, *nsxSubnetPort.Path, "RealizedLogicalPort"); err != nil {
+		log.Error(err, "failed to get realized status", "subnetport path", *nsxSubnetPort.Path)
+		if realizestate.IsRealizeStateError(err) {
+			log.Error(err, "the created subnet port is in error realization state, cleaning the resource", "subnetport", obj.UID)
+			// only recreate subnet port on RealizationErrorStateError.
+			if err := service.DeleteSubnetPort(obj.UID); err != nil {
+				log.Error(err, "cleanup error subnetport failed", "subnetport", obj.UID)
+				return err
+			}
+		}
+		return err
+	}
+	// TODO: avoid to get subnetport state again if we already got it.
+	nsxPortState, err := service.GetSubnetPortState(obj)
+	if err != nil {
+		return err
+	}
+	log.Info("Got the NSX subnet port state", "nsxPortState.RealizedBindings", nsxPortState.RealizedBindings)
+
+	gateway, netmask, err := service.getGatewayNetmaskForSubnetPort(obj)
+	if err != nil {
+		return err
+	}
+	ipAddress := v1alpha1.SubnetPortIPAddress{
+		Gateway: gateway,
+		IP:      *nsxPortState.RealizedBindings[0].Binding.IpAddress,
+		Netmask: netmask,
+	}
+	obj.Status.IPAddresses = []v1alpha1.SubnetPortIPAddress{ipAddress}
+	obj.Status.MACAddress = strings.Trim(*nsxPortState.RealizedBindings[0].Binding.MacAddress, "\"")
+	obj.Status.VIFID = *nsxPortState.Attachment.Id
+	// TODO: get the LSP ID when the subnet implementation is ready.
+	obj.Status.LogicalSwitchID = "b0b6a38b-942f-4029-8f2e-9987ad613197(fake)"
+
+	return nil
+}
+
+func (service *SubnetPortService) GetSubnetPathForSubnetPort(obj *v1alpha1.SubnetPort) (string, error) {
+	// TODO: For now, only the parent subnet is supported.
+	path := ""
+	if len(obj.Spec.Subnet) > 0 {
+		subnet := &v1alpha1.Subnet{}
+		namespacedName := types.NamespacedName{
+			Name:      obj.Spec.Subnet,
+			Namespace: obj.Namespace,
+		}
+		// TODO: Get the subnet's NSX resource path from subnet store after it is implemented.
+		if err := service.Client.Get(context.Background(), namespacedName, subnet); err != nil {
+			log.Error(err, "subnet CR not found", "subnet CR", namespacedName)
+			return path, err
+		}
+		path = subnet.Status.NSXResourcePath
+		if len(path) == 0 {
+			err := errors.New("empty NSX resource path from subnet")
```

Comment:
could you also add obj.Spec.Subnet in the error message?

Reply:
done